Support serializing longs to/from the local registry

jasonmalinowski opened the pull request:
Fixes dotnet/roslyn#19629.

heejaechang commented:
👍

heejaechang commented:
so will this be checked in instead of @CyrusNajmabadi 's fix?

jasonmalinowski commented:
We'll take @CyrusNajmabadi's fix since it should ensure we don't have _other_ things broken behind this. From there we'll test and then take my change.

CyrusNajmabadi commented:
👍 

Any thoughts on making our code natively support DateTime?

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs:
```diff
@@ -75,6 +75,27 @@ bool IOptionPersister.TryFetch(OptionKey optionKey, out object value)
+                    {
+                        object untypedValue = subKey.GetValue(key, defaultValue: optionKey.Option.DefaultValue);
+
+                        if (untypedValue is string)
```
pattern matching!

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs:
```diff
@@ -75,6 +75,27 @@ bool IOptionPersister.TryFetch(OptionKey optionKey, out object value)
+                            value = longValue;
+                            return suceeded;
+                        }
+                        else if (untypedValue is long)
```
pattern matching!

CyrusNajmabadi commented:
I'm fine with taking jasons.  right now jenkins is failing on my change.  so jason's may pass first.  IN that case, just send his in.

jinujoseph commented:
do you want to revert @CyrusNajmabadi's fix along with this PR ?

jasonmalinowski commented:
To be clear: I haven't even tried compiling my change yet. :smile:

jasonmalinowski commented:
OK, I've reverted @CyrusNajmabadi's change in this PR and have tested that I'm seeing it correctly be saved and restored as a QWORD from the registry.

I've rebased this PR so it can be targeted against either master or dev15.3-preview2. @jinujoseph which release would you like this to target?

jinujoseph commented:
Thanks Jason, master pls (dev15.3 final release)

jasonmalinowski commented:
Tagging @MattGertz for ask mode then.

This reenables the A/B test that was merged earlier this week and subsequently disabled due to a crash it was triggering somewhere else.

**Risk**: very low. Just adding a few special codepaths that aren't taken in any existing place.

**Performance impact**: none. Just a type check being added.

**Is this a regression from a previous update?** New code was added that was trying to read/write longs as part of an A/B test implementation. This triggered an underlying bug.

**Root cause analysis:** When you call RegistryKey.SetValue passing it an Int64, the framework does not store it as a QWORD but instead as a string. When we tried to deserialize this later, it failed because we expected to be getting an Int64 back. We never hit this bug as we never used to serialize longs to the registry before.

Why did we not see it in tests? Since once we've saved it to the registry we don't try to read the key again, we wouldn't have observed the problem until the _next_ run of VS. So integration tests would pass, but a _later_ run with the same code would then fail. When it was still in the PR, we would have seen the failure if by luck we ran the test on the same VM twice, but with our pool (that is continually rebuilt) it was improbable until it was already merged.

**How was the bug found?** Internal automated testing.

heejaechang commented:
@jasonmalinowski I get same issue with this as well

//
        public static readonly Option<uint> UsedSuggestedActionCount = new Option<** uint **>(nameof(AnalyzerABTestOptions), nameof(UsedSuggestedActionCount),
            defaultValue: 0, storageLocations: new LocalUserProfileStorageLocation(LocalRegistryPath + nameof(UsedSuggestedActionCount))
//

looks like uint require same special treatment as well.

CyrusNajmabadi commented:
We should just change that code to use an int to store.  There's no need for it to be uint.

heejaechang commented:
@CyrusNajmabadi sure we can do so. but right now, option serializer is a bit too fragile. I would expected it to actually handle all primitive types.

CyrusNajmabadi commented:
> I would expected it to actually handle all primitive types.

That's definitely out of scope for 15.3 :)

heejaechang commented:
@CyrusNajmabadi another issue is it will crash for ones which already has saved the value as string

CyrusNajmabadi commented:
That's solvable given that this hasn't made it out to the wild.  Internal roslyn dogfooders can just delete those regkeys locally.

heejaechang commented:
@jasonmalinowski @jinujoseph so I did the testing and it is failing on other parts. I feel a bit weird since I did same thing several days ago and it didn't happen. so I kind of wondering whether this might rather an issue with virtual registry key thing vs use?

jasonmalinowski commented:
@heejaechang Conversion from legacy string to the real value again is implemented in the PR.

So I gave a quick shot at supporting uint and given the .NET framework APIs presume you're working with signed ints, I don't trust myself to not write more bugs. If you really need it give it a shot, otherwise let's just use int.

heejaechang commented:
@jasonmalinowski I am fine changing it to int. I tried friday to push that to this PR and it got rejected. so I thought you would just put that change in this PR.

http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Experimentation/AnalyzerABTestOptions.cs,16

just change this to int from uint.

jasonmalinowski commented:
@heejaechang Not sure why you couldn't do the push, but I've made the change myself. Are we confident there are no other issues at this point? (FYI to @jinujoseph)

heejaechang commented:
@jasonmalinowski I am testing

heejaechang commented:
👍 just tested. everything working fine.
